src/models/mst.test.ts

```diff
@@ -299,4 +300,163 @@ describe("mst", () => {
+      name: types.string
+    })
+    .actions(self => ({
+      setName(name: string) { self.name = name; }
+    }));
+
+    const TodoList = types.model({
+      todos: types.array(Todo)
+    })
+    .actions(self => ({
+      updateAllTodosWithManager(_manager: any) {
+        _manager.managerUpdateAllTodos(self);
+      },
+      updateAllTodos() {
+        self.todos.forEach(todo => {
+          todo.setName("new list name");
+        });
+      }
+    }));
+
+    const MSTTodoManager = types.model({
+    })
+    .actions(self => ({
+      managerUpdateAllTodos(list: Instance<typeof TodoList>) {
+        list.todos.forEach(todo => {
+          todo.setName("new manager name");
+        });
+      }
+    }));
+
+    class MobXTodoManager {
+      constructor() {
+        makeObservable(this, {
+          managerUpdateAllTodos: action
+        });
+      }
+      
+      managerUpdateAllTodos(list: Instance<typeof TodoList>) {
+        list.todos.forEach(todo => {
+          todo.setName("new manager name");
+        });
+      }
+    }
+
+    // Main MST Tree
+    const todoList = TodoList.create({
+      todos: [
+        { name: "todo 1" },
+        { name: "todo 2" }
+      ]
+    });
+
+    // Secondary MST Tree
+    const mstManager = MSTTodoManager.create();    
+
+    // Manager that is not using MST and actions
+    const plainManager = {
+      managerUpdateAllTodos(list: Instance<typeof TodoList>) {
+        list.todos.forEach(todo => {
+          todo.setName("new manager name");
+        });
+      }
+    };
+
+    // MobX Manager, so it could still use observability, but because
+    // it isn't MST it doesn't mess up the middleware
+    const mobxManager = new MobXTodoManager();
+
+    const started : string[] = [];
+    const finished : string[] = [];
+    let topLevelCallId = 0;
+    const middleware = createActionTrackingMiddleware2({
+      filter(call) {
+        // record all calls even if they already have an environment
+        return true;
+      },
+      onStart(call) {
+        if (!call.env) {
+          call.env = { id: topLevelCallId };
+          topLevelCallId++;
+        }
+        started.push(`${call.name}:${call.env.id}`);
+      },
+      onFinish(call) {   
+        finished.push(`${call.name}:${call.env.id}`);     
+      }
+    });
+    const disposer = addMiddleware(todoList, middleware, true);
+
+    // Run an action in the Main Tree that does not pass through the secondary tree.
+    todoList.updateAllTodos();
+
+    expect(started).toEqual([
+      "updateAllTodos:0",
+      "setName:0",
+      "setName:0"
+    ]);
+
+    // reset the tracking props
+    started.length = 0;
+    finished.length = 0;
+    topLevelCallId = 0;
+
+    // Run an action in the Main Tree that passes through the secondary tree.
+    todoList.updateAllTodosWithManager(mstManager);
+
+    expect(started).toEqual([
+      "updateAllTodosWithManager:0",
+      "setName:1",
+      "setName:2"
+    ]);
```

Comment:
Add a comment here pointing out/explaining the incorrect result here. I had to read through the code a couple times to figure it out.

Reply:
I've refactored this to hopefully make it easier to understand.

Reply:
Very clear now